File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.js
/**
 * @param {number[][]} classes
 * @param {number} extraStudents
 * @return {number}
 */
var maxAverageRatio = function(classes, extraStudents) {
  const pq = new MaxPriorityQueue({ priority: x => x[0] });
  let sum = 0;
  
  for (let [pass, total] of classes) {
    sum += pass / total;
    pq.enqueue([profit(pass, total), pass, total]);
  }
  
  while (extraStudents--) {
    let [currProfit, pass, total] = pq.dequeue().element;
    sum += currProfit;
    pass++;
    total++;
    pq.enqueue([profit(pass, total), pass, total]);
  }
  
  return sum / classes.length;
};

const profit = (pass, total) => (pass + 1) / (total + 1) - pass / total;

File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        heap = [(a / b - (a+1) / (b+1), a, b) for a, b in classes]
        heapq.heapify(heap)
        while extraStudents:
            _, a, b = heapq.heappop(heap)
            a, b = a + 1, b + 1
            heapq.heappush(heap, (a / b - (a+1) / (b+1), a, b))
            extraStudents -= 1
        return sum(a / b for _, a, b in heap) / len(heap)